### PR TITLE
Allow for `/boot` to be customized

### PR DESCRIPTION
**Pull request by gicmo**
The interesting commit is the last one, with the commit message:

Since the LVM support was added to all distros, our disk related code is adaptive, i.e. we will set the correct BLS and grub2 prefix if there a `boot` partiton is present in the layout after all customizations happen, which includes LVMification. One thing that was not yet fully working was layouts that do not yet have a `/boot` partition but allow LVMification. In that case `NewPartitionTable` and if `/boot` was the first (or only) customization, would LVMify the partition which in turn would create the `/boot` partition; but after `newPT.ensureLVM()` the call to `newPT.createFilesystem` with `/boot` would try to create another `/boot` mountpoint. In order to deal with this situation correctly we are now using a two phase approach: 1) enlarge existing mountpoints and collect new ones. 2) if there are new ones and LMVify was allowed, switch to LVM layout. Do a second pass and now create or enlarge existing partitions, handling `/boot` in the process.

### CODE REVIEW

**Review by teg — DISMISSED** [review]

File: internal/disk/disk.go [hunk 1]
```diff
@@ -164,3 +168,40 @@ func NewVolIDFromRand(r *rand.Rand) string {
 	}
 	return hex.EncodeToString(volid)
 }
+
+// IsMountpointAllowed checks if the mountpoint is clean and allowed
+func IsMountpointAllowed(mountpoint string, allowlist []string) bool {
```
[the remark this answers is not in the document]
Comment (gicmo) [reply internal/disk/disk.go]: Should be fixed.

**Comment by gicmo** [comment]
I have added a simple trie data structure implementation and converted the mount point allow list to it in order to allow `/boot` but none if its sub-directories. Additionally, the customisation code is now extracted into a new helper method `applyCustomizations` and re-used between the two passes.

**Review by teg — DISMISSED** [review]
Nice stuff! I like the trie

File: internal/disk/path_trie.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+func (trie *PathTrie) get(path []string) (*PathTrie, []string) {
+	if len(path) < 1 {
+		panic("programming error: expected root node")
+	}
+
+	var node *PathTrie
+	var prefix int
+	for i := range trie.Paths {
+		if prefix = trie.Paths[i].match(path); prefix != 0 {
```
Comment (thozza) [inline internal/disk/path_trie.go:60]: If the `trie.Paths[i]` contains more than one directory name in `Name`, then it is not enough to just check that the returned `prefix` is non-zero. If the `trie.Paths[i]` `Name` contains e.g. `[]string{"lib", "osbuild"}` and the `path` is `[]string{"lib", "chrony"}`, they you'll get non-zero `prefix`, but I would say that it is not a match. The `prefix` should not be really lower than the node's `len(Name)` to call it a match.

File: internal/disk/path_trie.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	path = strings.Trim(path, "/")
+	if path == "" {
+		return []string{}
+	}
+
+	return strings.Split(path, "/")
+}
+
+type PathTrie struct {
+	Name    []string
```
[the remark this answers is not in the document]
Comment (thozza) [reply internal/disk/path_trie.go:18]: For:
```go
map[string]interface{}{
	"/":                            common.IntToPtr(1),
	"/var":                         common.IntToPtr(2),
	"/var/lib/chrony":              common.IntToPtr(7),
	"/var/lib/osbuild":             common.IntToPtr(3),
	"/var/lib/osbuild/store/cache": common.IntToPtr(4),
	"/boot":                        common.IntToPtr(5),
	"/boot/efi":                    common.IntToPtr(6),
}
```
you'll get
```go
PathTrie{
	Name:    []string{},
	Payload: common.IntToPtr(1),
	Paths: []*PathTrie{
		{
			Name:    []string{"boot"},
			Payload: common.IntToPtr(5),
			Paths: []*PathTrie{
				{
					Name:    []string{"efi"},
					Payload: common.IntToPtr(6),
				},
			},
		},
		{
			Name:    []string{"var"},
			Payload: common.IntToPtr(2),
			Paths: []*PathTrie{
				{
					Name:    []string{"lib", "chrony"},
					Payload: common.IntToPtr(7),
					Paths: []*PathTrie{
						{
							Name:    []string{"osbuild"},
							Payload: common.IntToPtr(3),
							Paths: []*PathTrie{
								{
									Name:    []string{"store", "cache"},
									Payload: common.IntToPtr(4),
								},
							},
						},
					},
				},
			},
		},
	},
}
```
which does not look correct.

**Review by thozza — COMMENTED** [review]
Great work! The reworked mount point policies and introduction of `PathPolicy` structure is just great 😍 

However I think that I found an issue with the `PathTrie` implementation. In the way you use it with `PathPolicy` in unit tests and in composer's implementation, it will behave as expected. However there is a corner case, in which it will produce unexpected results. Specifically if the map used to instantiate the Trie skips one directory level in the paths that it defines (e.g. is defines `/var/`,  `/var/lib/osbuild` and `/var/lib/chrony`, but no `/var/lib`). This is not obvious, nor documented (as far as I checked) and may lead to unintentional coding error.

File: internal/disk/path_trie.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gicmo) [reply internal/disk/path_trie.go:18]: > Is there any particular reason why you didn't choose to implement the trie with a single directory being represented by a single node?

Ironically, that was the first version but I didn't like the un-needed intermediate nodes; and it made it impossible to distinguish between "helper" nodes and actual nodes. This is currently not important but I had a version with a better error message (with the help of a `fullname` field) that would show you which exact policy is preventing the mountpoint.

[the remark this answers is not in the document]
Comment (gicmo) [reply internal/disk/path_trie.go:18]: > However there is a corner case, in which it will produce unexpected results.

Yeah, I thought we would not need that special case, but you are of course right that it should at least be documented. On a second thought, maybe I should just fix the implementation because it is weird to have that limitation even though it will make the code more complicated.

[the remark this answers is not in the document]
Comment (thozza) [reply internal/disk/path_trie.go:18]: Reimplementation and preventing hitting this case sounds better to me than documenting this "feature". Thanks a lot!

**Review by thozza — APPROVED** [review]
Everything looks amazing 🚀
